Sources/NIOHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,333 @@
+                return self.eventLoopGroup.next().makeSucceededFuture(())
+            }
+            let address = HTTPClient.resolveAddress(host: request.host, port: request.port, proxy: configuration.proxy)
+            return bootstrap.connect(host: address.host, port: address.port).flatMap { channel in
+                let http1PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<Void> = { pipeline in
+                    pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes).map { _ in
+                        let http1Provider = ConnectionProvider.http1(HTTP1ConnectionProvider(group: self.eventLoopGroup, key: key, configuration: self.configuration, initialConnection: Connection(key: key, channel: channel)))
+                        self[key] = http1Provider
+                        http1Provider.closeFuture.whenSuccess {
+                            self[key] = nil
+                        }
+                        protocolChoosenPromise.succeed(http1Provider)
+                    }
+                }
+                
+                let h2PipelineConfigurator: (ChannelPipeline) -> EventLoopFuture<Void> = { pipeline in
+                    channel.configureHTTP2Pipeline(mode: .client).map { multiplexer in
+                        let http2Provider = ConnectionProvider.http2(HTTP2ConnectionProvider(group: self.eventLoopGroup, key: key, configuration: self.configuration, channel: channel, multiplexer: multiplexer))
+                        self[key] = http2Provider
+                        http2Provider.closeFuture.whenSuccess {
+                            self[key] = nil
+                        }
+                        protocolChoosenPromise.succeed(http2Provider)
+                    }
+                }
+                
+                if request.useTLS {
+                    _ = channel.pipeline.addSSLHandlerIfNeeded(for: request, tlsConfiguration: self.configuration.tlsConfiguration).flatMap { _ in
+                        channel.pipeline.configureHTTP2SecureUpgrade(h2PipelineConfigurator: h2PipelineConfigurator, http1PipelineConfigurator: http1PipelineConfigurator)
+                    }
+                } else {
+                    _ = http1PipelineConfigurator(channel.pipeline)
+                }
+                
+                return protocolChoosenPromise.futureResult.flatMap { provider in
+                    provider.getConnection()
+                }
+            }
+        }
+    }
+
+    func release(_ connection: Connection) {
+        if let provider = self[connection.key] {
+            provider.release(connection: connection)
+        } else {
+            //FIXME: Is this acceptable to discard this ELF?
+            _ = connection.channel.close()
+        }
+    }
+
+    fileprivate struct Key: Hashable {
+        
+        init(url: URL) throws {
```

Comment:
why not use `HTTPClient.Request` here? It already has all necessary fields resolved and verified?

Reply:
That's because `Key` is used in contexts where the whole Request wouldn't seem really appropriate, for instance all `Connection` objects have a `key: Key` property, but they aren't linked to any specific `Request`, same thing for the connection providers.